### ISSUE TITLE
pinstall.sh apears not to be executed in 2createpackages but doinst.sh is for slackware packages

### ISSUE THREAD

**Issue by s243a** [puppylinux-woof-CE/woof-CE#1789]
I was writing a pinstall.sh script, and I was wondering if I could detect the context in the script (e.g. woof, vs package_chooser.sh vs petget) and when I search the github code it appears to me that pinstall.sh is not executed in woof-CE but the post install scripts are executed for slackware packages (see [/woof-code/2createpackages#L309](https://github.com/puppylinux-woof-CE/woof-CE/blob/7cfdaea8673ba60ca71e5ce783f599a2a0e18bae/woof-code/2createpackages#L309)). 

Am I missing something or is this intentional? Is there a reason that we want to execute the post install scripts for slackware packages but not for .pet or .deb? It's possible I'm reading the code here wrong but I don't think so.

**Comment by peabee** [comment]
All the pet pinstalls are combined into a single large pinstall that is executed in ./3builddistro

You can see it in sandbox3

**Comment by 01micko** [comment]
Just make sure you use _relative_ path.

**Comment by peabee** [comment]
There must also be no "exit" in the pinstall otherwise the combined pinstall will terminate prematurely

**Comment by s243a** [comment]
Thankyou for the tips 01micko and peabee. :) Regarding relative paths, I'm wondering if I can do a chroot. If the pinstall scripts are executed in the order as given by DISTRO_PKGS_SPECS-* then I can make sure the order implies that anything that I need for my script will be in place or perhaps the order is determined by dependencies? 

Anyway, being able to do a chroot would be very helpful because puppy isn't a prefixed system like termux.

**Comment by s243a** [comment]
> Thankyou for the tips 01micko and peabee. :) Regarding relative paths, I'm wondering if I can do a chroot. If the pinstall scripts are executed in the order as given by DISTRO_PKGS_SPECS-* then I can make sure the order implies that anything that I need for my script will be in place or perhaps the order is determined by dependencies?
> 
> Anyway, being able to do a chroot would be very helpful because puppy isn't a prefixed system like termux.

Aside from whether or not a chroot would work in a pinstall script for woof-CE, it looks like it might be problematic in PUPMODE=13 on a running puppy since the current directory in pupmode=13 is:
```
DIRECTSAVEPATH="/initrd${SAVE_LAYER}" #SAVE_LAYER is in /etc/rc.d/PUPSTATE
```
[/usr/local/petget/installpkg.sh#L216](https://github.com/puppylinux-woof-CE/woof-CE/blob/7cfdaea8673ba60ca71e5ce783f599a2a0e18bae/woof-code/rootfs-skeleton/usr/local/petget/installpkg.sh#L216)

and then the script is evaluated here:
```
#post-install script?...
#          puppy         slackware       debian/ubuntu/etc
for i in pinstall.sh install/doinst.sh DEBIAN/postinst
do
	[ ! -e "$DIRECTSAVEPATH/$i" ] && continue
	cd $DIRECTSAVEPATH/
	LANG=$LANG_USER sh ${i}
	sync
	rm -f ${i}
done
```
[/usr/local/petget/installpkg.sh#L377](https://github.com/puppylinux-woof-CE/woof-CE/blob/7cfdaea8673ba60ca71e5ce783f599a2a0e18bae/woof-code/rootfs-skeleton/usr/local/petget/installpkg.sh#L377)

Therefore, if one is writing a pinstall.sh script they might want to first check if PUPMOE=13 and if so then use absolute paths and of course in such a scenario don't execute commands in a chroot. 

In the pinstall.sh script we could test for it like this:
```
[[ "`pwd`" =~ .*initrd.* ]] && echo "use absolute paths
```

**Comment by 01micko** [comment]
The whole point of relative paths is so that they work in a `chroot`.

If there is a problem installing in `PUPMODE=13` then it is an issue of the pupmode.

I remember quite awhile ago we decided to install in 13 to `/` but that may have been reverted. I'll have to search commits, when I can, maybe over the weekend.

BTW, best not to use bashism in a pinstall. Barry's method was rather simple -

```sh
#!/bin/sh
if [ `pwd` = '/' ] ; then
  echo 'running os' # pinstall is at '/'
else
  echo 'chroot' # obviously some path to the root tree of the target
fi
```
However, if there _is_ a problem in 13 with rel paths then slack/deb packages may not be installing properly.

The question is, both with `pkg` and `ppm` .. "How to handle this?". I think installing to `/` is that answer, with appropriate free space checks.

**Comment by jamesbond3142** [comment]
As @micko @peebee said. One should not depend on the absolute location of the installation path; it should be taken that the ```current directory``` is the root of the installation; and use relative path for _everything_. This is the only way that the install script would work when run under petget or under woof-CE build; and it is a sane way to do it (Slackware does the same thing, and debian packages too as far as I know).

I don't understand __why__ you need to detect who run the install script and/or the installation path etc. If you really really need to be that specific, then in your install script do ```chroot . /path/to/your/special/script/in/the/chroot/special-script.sh``` and there you go, your "special-script.sh" will always have "/" as the installation directory.

On a tangent note, install script should be as simple as possible. It's even better if they don't exist at all. Don't over-complicate things. Only do what is necessary for that particular package you are installing and nothing else. Do not change system settings unless it is absolutely necessary (e.g. inserting a startup script is ok, modifying other package's startup script is __not ok__). Do not attempt to resolve dependencies or pull extra packages - these are the job of package managers, not install script (if the package manager doesn't do it - the fix the package manager, don't do it in install script). Do not install additional un-tracked files (if it does, then package manager can't remove it - it becomes the job of the uninstall script to do so). Note some that some package manager aren't recursive-safe, if you attempt to install dependencies via package manager (with the hope that the installed dependencies be tracked) while you're running inside install script under an already running package manager, the package manager may fail in unexpected way.

Of course there is an exception for every rule, but this rule should not be broken: if you are going to break these rule of thumb, at the very least __alert and get permission from the user and give the chance to abort__ before you're doing it. Because I as a user for example don't welcome an intrusive install script that does a lot of things behind my back, and would consider such script as malicious.

An install script purpose only this and only this: to configure the package that contains it so that it can run properly. "configure" here means updating the package/system that could not otherwise be done during build/packaging time. E.g. running ```update-mime-database``` or ```chown```-ing the certain files/directories in the package to known uid/gid.

**Comment by Butterfly** [comment]
> The whole point of relative paths is so that they work in a `chroot`.

Yeehaw on that! I debootstrap Debian on regular occasion. Used to have A LOT of problems where my HOST system would get botched in the process. Couple years ago, I consciously noticed where the following was how the base `debootstrap` followed by installing `linux-image` was put in place e.g.:

```
vmlinuz -> boot/vmlinuz-4.13.0-36-generic
initrd.img -> /boot/initrd.img-4.13.0-36-generic
```

There was a size difference or something that was the ah-ha moment, but I can't recreate what happened now. The outcome back then was that it was pointing an **absolute** `initrd.img` link to my **HOST** system, *NOT* the `chroot` environment.

Problem? Seems like when upgrades were done under `chroot`, that might just be when my HOST system started getting botched. I'm presuming maybe the HOST system was what was being upgraded, NOT the GUEST.

NB: No, it wasn't me as the User creating that circumstance. That was the default during installation back then. These days BOTH `vmlinuz` and `initrd.img` are relative links. #ThankYou, Developers!! Cross-contamination has been basically non-existent in the last couple years. I don't think it's a coincidence. :smile:

**Comment by Butterfly** [comment]
The importance of considering `ln -s` symlinks versus `mount --bind`/`mount -B` mount points was something else I learned the very hard way. How I learned that was...

I'm on dialup. Because of that, I hoard all my downloads to save me time as well as to decrease stress on host repository servers AND my local Internet provider service. One day I got the bright idea to symlink my massive .deb archives hoard to the /chroot/var/cache/apt/archives directory before beginning a `debootstrap`.

The entire download session *appeared* to process as usual [0]. When I chrooted in, my root user started yelling, "**I HAVE NO NAME!**" Everything failed within those affected `chroot` environments from that point on.

That went on for EIGHT MONTHS where I couldn't do ANYTHING during those attempts. Somewhere somehow the timing was that I grasped that `mount -B` MIGHT work for what I was doing so I tried that. **BINGO!**

One of my personal longstanding TO-DO list items is to finally track down the technical explanation for why `debootstrap` would *appear* to work 100% successfully with that (absolute) symlink but obviously did not because of root's resulting identify crisis. That identity crisis was instantaneously corrected with `mount -B` as the alternative.

[0] There's a very short little `debootstrap` message that appears if things did *NOT* go quite as planned. I can't remember what that message is, but it can be missed if one is in a hurry. It kind of "blends in" with the rest of the terminal feedback right at the very end.
